lib/matplotlib/layout_engine.py

```diff
@@ -30,29 +30,39 @@ class LayoutEngine:
     Base class for Matplotlib layout engines.
 
     A layout engine can be passed to a figure at instantiation or at any time
-    with `~.figure.Figure.set_layout_engine`.  However, note note that layout
-    engines affect the creation of colorbars, so
-    `~.figure.Figure.set_layout_engine` should be called before any colorbars
-    are created.
-
-    Once attached to a figure, the layout engine ``execute`` function is called
-    at draw time by `~.figure.Figure.draw`, providing a special draw-time hook.
-
-    Currently, there are two properties of ``LayoutEngine`` classes that are
-    consulted while manipulating the figure.  ``engine.colorbar_gridspec``
-    tells `.Figure.colorbar` whether to make the axes using the gridspec
-    method (see `.colorbar.make_axes_gridspec`) or not
-    (see `.colorbar.make_axes`); `.ConstrainedLayoutEngine` sets this to
-    *False*, `.TightLayoutEngine` to *True*.  The second property is
-    ``engine.adjust_compatible`` that stops `.Figure.subplots_adjust` from
-    being run if it is not compatible with the layout engine
-    (`.ConstrainedLayoutEngine` sets this to *False* also).
+    with `~.figure.Figure.set_layout_engine`.  Once attached to a figure, the
+    layout engine ``execute`` function is called at draw time by
+    `~.figure.Figure.draw`, providing a special draw-time hook.
+
+    .. note ::
+
+       However, note that layout engines affect the creation of colorbars, so
+       `~.figure.Figure.set_layout_engine` should be called before any
+       colorbars are created.
+
+    Currently, there are two properties of `LayoutEngine` classes that are
+    consulted while manipulating the figure:
+
+    - ``engine.colorbar_gridspec`` tells `.Figure.colorbar` whether to make the
+       axes using the gridspec method (see `.colorbar.make_axes_gridspec`) or
+       not (see `.colorbar.make_axes`);
+    - ``engine.adjust_compatible`` stops `.Figure.subplots_adjust` from being run
+       if it is not compatible with the layout engine.
+
+    To implement a custom `LayoutEngine`:
+
+    1. override ``_adjust_compatible`` and ``_colorbar_gridspec``
+    2. override `LayoutEngine.set` to update *self._params*
+    3. override `LayoutEngine.execute` with your implementation
+
     """
+    # override these is sub-class
+    _adjust_compatible = None
```

Comment:
Just for my education, why are these better defined here than in the __init__?

Reply:
It is a couple of marginal things for me.  

Defining them in `__init__` sub-classes will either have to be sure to call `__super__()` (which they do have to here for other reasons so this is a bit of a wash here) or remember to define these values directly.  By putting them on the class object you have taken a bit of a burden / defused a trap for the down stream user.  In the case of a user-implemented subclass that does not want to make itself configurable at init time (because it has no configuration or just because), then  the do not even have to define an `__init__` with these values as class level attributes.

Another advantage of this is because these really are class-constants (at least for these two) in the sense that these should _never_ change at run time.  Putting them here indicates that and makes it easy to, when skimming the code, see these right away as different "class state things" not "instance state things".  It is a small thing, but it gives a hint to someone reading this to implement their own.

By making these default to `None` (and checking it in the properties!) you force a down-stream implementer to think about this once to make a choice.

Reply:
OK, thanks for the insight!